source/particle/world.cc

```diff
@@ -206,6 +210,88 @@ namespace aspect
+                  {
+                    for (unsigned int i = particles_in_cell; i < min_particles_per_cell; ++i)
+                      {
+                        std::pair<aspect::Particle::types::LevelInd,Particle<dim> > new_particle = generator->generate_particle(cell,0);
+
+                        Vector<double> value(this->introspection().n_components);
+                        std::vector<Tensor<1,dim> > gradient (this->introspection().n_components,Tensor<1,dim>());
+
+                        std::vector<Vector<double> >  solution(1,value);
+                        std::vector<std::vector<Tensor<1,dim> > > gradients(1,gradient);
+
+                        std::vector<Point<dim> >     particle_points(1);
+                        particle_points[0] = this->get_mapping().transform_real_to_unit_cell(cell, new_particle.second.get_location());
+
+                        const Quadrature<dim> quadrature_formula(particle_points);
+                        FEValues<dim> fe_value (this->get_mapping(),
+                                                this->get_fe(),
+                                                quadrature_formula,
+                                                update_values |
+                                                update_gradients);
+
+                        fe_value.reinit (cell);
+                        fe_value.get_function_values (this->get_solution(),
+                                                      solution);
+                        fe_value.get_function_gradients (this->get_solution(),
+                                                         gradients);
+
+                        property_manager->initialize_late_particle(new_particle.second,
+                                                                   particles,
+                                                                   *interpolator,
+                                                                   solution[0],
+                                                                   gradients[0]);
+
+                        particles.insert(new_particle);
+                      }
+                  }
+
+                // Remove particles if necessary
+                if ((max_particles_per_cell > 0) && (particles_in_cell > max_particles_per_cell))
+                  {
+                    unsigned int particle_index = 0;
+                    const std::pair<typename std::multimap<types::LevelInd, Particle<dim> >::iterator, typename std::multimap<types::LevelInd, Particle<dim> >::iterator>
+                    particles_in_cell = particles.equal_range(found_cell);
+
+                    for (typename std::multimap<types::LevelInd, Particle<dim> >::iterator particle = particles_in_cell.first;
+                         particle != particles_in_cell.second; ++particle_index)
+                      {
+                        if (particle_index % GeometryInfo<dim>::max_children_per_cell != 0)
```

Comment:
The procedure was different. When I wrote this code I assumed that the maximum number of particles would exceed the maximum number when coarsening happens, so to keep the particle density constant I would need to reduce the number of particles by a factor of 2^dim (the maximum number of children of a cell). Now with the new code it is awkward to have particle removal work differently from particle creation, so I will rewrite this to only remove `particles_in_cell - max_particles_per_cell` particles. However, I will need to remove random particles from the cell, because just removing from the beginning or the end of the particle list might look awkward for ordered particle distributions (suddenly a part of the cell will be cleared of particles, and another part still has the old particle density).